state-chain/chains/src/btc/utxo_selection.rs

```diff
@@ -1,68 +1,126 @@
+///
+/// On success, the `(selected_utxos and accumulated_total)` is returned.
+/// On failure the appropriate error is returned, and `available_utxos` is still modified. It is
+/// expected that the user will roll back storage.
 ///
 /// In the case where the fee to spend the utxo is higher than the amount locked in the utxo, the
 /// algorithm will skip the selection of that utxo and will keep it in the list of available utxos
 /// for future use when the fee possibly comes down so that it is feasible to select these utxos.
 ///
-/// Note that on failure (when `None` is returned) this may still modify `available_utxos`, which
-/// is not a concern because the caller will ignore that value in that case.
+/// The algorithm for the utxo selection works as follows:
+/// 1. Filter out all available utxos whose fees are >= amount, they are excluded from the
+/// selection.
+///
+/// 2. Find a optimal range such that `sum(utxo[first ..= last]) >= target_amount`
+/// When the selected range is < selection_limit, move the `first` index to the left
+///
+/// e.g. using a selection limit of 3
+/// Currently 2 utxos are selected. Move the start index to the left.
+///                    | end
+/// 10 9 8 7 6 5 4 3 2 1
+///                |<| Start
+/// New end points to "1" and new start points to "3"
+///
+/// When the selection limit is reached, both the start and end pointer are moved.
+///                  |<| end
+/// 10 9 8 7 6 5 4 3 2 1
+///            |<| Start
+/// New end points to 2 and new start points to 5, keeping the selected utxos to the limit of 3.
+///
+/// In the worst case scenarios, the largest N utxos are selected.
+///
+/// If after this step, the accumulated amount is still below the target, report as failure.
+/// Thought this is extremely unlikely to happen since we proactively consolidate our utxos.
+///
+/// 3. We then take 1 more utxo if within limit, to actively reduce fragmentation.
+/// The skipped utxos are appended back to the `available_utxos` for future use.
 pub fn select_utxos_from_pool(
 	available_utxos: &mut Vec<Utxo>,
 	fee_info: &BitcoinFeeInfo,
 	amount_to_be_spent: u64,
-) -> Option<(Vec<Utxo>, u64)> {
-	if available_utxos.is_empty() {
-		return None
+	maybe_limit: Option<u32>,
+) -> Result<(Vec<Utxo>, u64), UtxoSelectionError> {
+	// If no selection limit is given, selecting all utxos is allowed.
+	let selection_limit = match maybe_limit {
+		Some(limit) => limit as usize,
+		None => available_utxos.len(),
+	};
+
+	// 1. Filter out utxos whose fees are too high.
+	let mut skipped_utxos = available_utxos
+		.extract_if(|utxo| utxo.amount <= fee_info.fee_for_utxo(utxo))
+		.collect::<Vec<_>>();
+
+	if available_utxos.is_empty() || selection_limit == 0usize {
+		return Err(UtxoSelectionError::NoUtxoAvailableAfterFeeReduction)
 	}
 
+	// 2. Find the optimal `first` and `last` index, such that
+	// `sum(utxo[first ..= last]) >= target_amount`
 	available_utxos.sort_by_key(|utxo| sp_std::cmp::Reverse(utxo.amount));
```

Comment:
If we don't use `Reverse` then the values will be in ascending order and the algorithm should be easier to write (since we won't have the confusing situation where the 'first' index is greater than the 'last').

Reply:
Btw. we should sort by the utxo's *net* value, not by utxo amount.